### PR TITLE
BUG: pivot_table losing tz

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This gets rid of the last values_from_object usage (pending other PRs)

`_tile_compat` likely makes more sense as an Index method.  I kept it here as a proof of concept because I think we actually need it in the EA interface too.  Being able to broadcast a length=1 EA to a length=N EA will be necessary for some of the arithmetic perf going on.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
cc @jorisvandenbossche @TomAugspurger @jreback thoughts on adding something like `tile` or `broadcast_to` to EA interface?

**Comment by TomAugspurger** [comment]
What are your proposed semantics of an EA.tile? If it's restricted to a single dimension, can to be emulated using `_concat_same_type`?

**Comment by jbrockmendel** [comment]
> What are your proposed semantics of an EA.tile? If it's restricted to a single dimension, can to be emulated using _concat_same_type?

The base class implementation would probably just use _concat_same_type.  For ndarray-backed EAs I'd use broadcast_to which doesn't make a copy, but results in a non-writeable array (in that sense it would behave more like broadcast_to than np.tile)

**Review by jreback — CHANGES_REQUESTED** [review]
I agree we should just have EA.tile; this is very messy otherwise.

**Comment by jreback** [comment]
can you see if we have any issues about this (pivot and tz), I think we might.

File: pandas/core/reshape/util.py [hunk 1]
```diff
@@ -51,9 +52,58 @@ def cartesian_product(X):
...
-            np.repeat(np.asarray(com.values_from_object(x)), b[i]), np.product(a[i])
-        )
-        for i, x in enumerate(X)
-    ]
+    return [_tile_compat(np.repeat(x, b[i]), np.product(a[i])) for i, x in enumerate(X)]
+
+
+def _broadcast_tile(arr: np.ndarray, num: int) -> np.ndarray:
+    """
+    Emulate np.tile but using views instead of copies.
```
Comment (jorisvandenbossche) [inline pandas/core/reshape/util.py]: Is this an important optimization? Currently we also use `np.tile` ?

**Comment by jorisvandenbossche** [comment]
> The base class implementation would probably just use _concat_same_type.

We could also simplify this a lot by using `_concat_same_type` here (which basically repeats my inline question: is this optimization important?)

Regarding tile in the EA interface: numpy also has no tile method, so it's not that it would make it more compatible with numpy arrays (for that, we would need to implement a protocol to make `np.tile` work on EAs). 
Personally, having the centralized "tile for EA function" here is fine for me.

File: pandas/core/reshape/util.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/reshape/util.py]: For large inputs yes

**Comment by jbrockmendel** [comment]
> We could also simplify this a lot by using _concat_same_type here (which basically repeats my inline question: is this optimization important?)

This is actually close to orthogonal to the broadcast_to optimization.  When I disable the broadcat_to optimization and just use np.tile, this implementation is about 2x faster than one that uses _concat_same_type (based on `arr = pd.date_range('2016-01-01', periods=10**5, freq='S')` and `num=5`)

**Comment by jbrockmendel** [comment]
> can you see if we have any issues about this (pivot and tz), I think we might.

searching on "pivot" didnt show anything tz-related

**Comment by jorisvandenbossche** [comment]
> This is actually close to orthogonal to the broadcast_to optimization. 

Why? It seems the reason you are not using the slower/simpler `np.tile` / `_concat_same_type` is to be able to use this optimization?

> this implementation is about 2x faster than one that uses _concat_same_type 

This is only for the tile? What does it give in the full pivot operation?

**Comment by jbrockmendel** [comment]
> Why? It seems the reason you are not using the slower/simpler np.tile / _concat_same_type is to be able to use this optimization?

No, its also because the non-concat_same_type implementation is 2x faster, even without the broadcast_to optimization.

For arithmetic I'm soon going to need to broadcast/tile length=1 EAs to length=N, which is very wasteful without the broadcast_to; that is the forward-looking motivation for using the broadcast_to optimization.

> This is only for the tile? What does it give in the full pivot operation?

No idea.

**Comment by jbrockmendel** [comment]
Updated to use a simpler implementation that does not do any optimizations.

**Comment by jreback** [comment]
looks fine, can you add a release note, ping on green.

**Comment by jbrockmendel** [comment]
ping

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thxs

**Comment by jbrockmendel** [comment]
booyah!  there goes our last usage of values_from_object

**Pull request merged**